server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java

```diff
@@ -0,0 +1,148 @@
+
+    public TranslogTransferManager(
+        TransferService transferService,
+        BlobPath remoteBaseTransferPath,
+        FileTransferListener fileTransferListener,
+        UnaryOperator<Set<TransferFileSnapshot>> exclusionFilter
+    ) {
+        this.transferService = transferService;
+        this.remoteBaseTransferPath = remoteBaseTransferPath;
+        this.fileTransferListener = fileTransferListener;
+        this.exclusionFilter = exclusionFilter;
+    }
+
+    public boolean transferSnapshot(TransferSnapshot transferSnapshot, TranslogTransferListener translogTransferListener)
+        throws IOException {
+        List<Exception> exceptionList = new ArrayList<>(transferSnapshot.getTranslogTransferMetadata().getCount());
+        Set<TransferFileSnapshot> toUpload = new HashSet<>(transferSnapshot.getTranslogTransferMetadata().getCount());
+        try {
+            toUpload.addAll(exclusionFilter.apply(transferSnapshot.getTranslogFileSnapshots()));
+            toUpload.addAll(exclusionFilter.apply(transferSnapshot.getCheckpointFileSnapshots()));
+            final CountDownLatch latch = new CountDownLatch(toUpload.size());
+            LatchedActionListener<TransferFileSnapshot> latchedActionListener = new LatchedActionListener<>(
+                ActionListener.wrap(fileTransferListener::onSuccess, ex -> {
+                    assert ex instanceof FileTransferException;
+                    logger.error(
+                        () -> new ParameterizedMessage(
+                            "Exception during transfer for file {}",
+                            ((FileTransferException) ex).getFileSnapshot().getName()
+                        ),
+                        ex
+                    );
+                    FileTransferException e = (FileTransferException) ex;
+                    fileTransferListener.onFailure(e.getFileSnapshot(), ex);
+                    exceptionList.add(ex);
+                }),
+                latch
+            );
+            toUpload.forEach(
+                fileSnapshot -> transferService.uploadBlobAsync(
+                    fileSnapshot,
+                    remoteBaseTransferPath.add(String.valueOf(fileSnapshot.getPrimaryTerm())),
+                    latchedActionListener
+                )
+            );
+            try {
+                if (latch.await(TRANSFER_TIMEOUT_IN_MILLIS, TimeUnit.MILLISECONDS) == false) {
+                    Exception ex = new TimeoutException("Timed out waiting for transfer of snapshot " + transferSnapshot + " to complete");
+                    exceptionList.forEach(ex::addSuppressed);
+                    throw ex;
+                }
+            } catch (InterruptedException ex) {
+                exceptionList.forEach(ex::addSuppressed);
+                logger.error(() -> new ParameterizedMessage("Time failed for snapshot {}", transferSnapshot), ex);
+                Thread.currentThread().interrupt();
+                throw ex;
```

Comment:
Is this catch block necessary given that it re-throws and is caught below? I think exceptionList ends up being unused. Also I think the interrupt status for the thread will be unset in the catch block below. It also looks like this error message will be logged twice if I'm reading this correctly.

Reply:
exceptionlist is being added to suppressed as a part of ex, which is what gets propagated, so exceptionlist is being consmed. Yes the exception will be logged twice but messages would be different, will see if we can have just one log line